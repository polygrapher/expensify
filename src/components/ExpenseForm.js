import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const expense = props.expense || {};
        const {description = '', note = '', amount, createdAt} = expense;

        this.state = {
            error: '',
            description,
            note,
            amount: amount ? (amount / 100).toString() : '',
            createdAt: createdAt ? moment(createdAt) : moment(),
            calendarFocused: false
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}));
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    }

    onSubmit = (e) => {
        e.preventDefault();
        const {amount, description, createdAt, note} = this.state;

        const error = !amount || !description ? 'Please provide amount and description' : '';
        if (!error) {
            this.props.onSubmit({
                description,
                amount: parseFloat(amount, 10) * 100,
                createdAt: createdAt.valueOf(),
                note
            })
        } else {
            this.setState(() => ({error}));
        }
    }

    render() {
        const {error, description, note, createdAt, amount} = this.state;
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {error && <p className="form__error">Error: {error}</p>}
                <input
                    className="text-input"
                    type="text"
                    placeholder="Description"
                    autoFocus
                    onChange={this.onDescriptionChange}
                    value={this.state.description}
                />
                <input
                className="text-input"
                    type="text"
                    placeholder="Amount"
                    onChange={this.onAmountChange}
                    value={amount}
                />
                <SingleDatePicker
                    date={createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="textarea"
                    placeholder="Add a note for your description"
                    onChange={this.onNoteChange}
                    value={note}
                >
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}