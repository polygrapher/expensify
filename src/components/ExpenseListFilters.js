import React from 'react';
import {connect} from 'react-redux';
import { bindActionCreators } from 'redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input type="text" className="text-input" placeholder="Search Expenses" value={this.props.filters.text} onChange={this.onTextChange}/>
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Value</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                            startDateId="expenseFiltersStartDate" // PropTypes.string.isRequired,
                            endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                            endDateId="expenseFiltersEndDate" // PropTypes.string.isRequired,
                            onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                            focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = ({filters}) => (
    {filters}
);

const mapDispatchToProps = ((dispatch) => bindActionCreators({
    setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate
}, dispatch));

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
