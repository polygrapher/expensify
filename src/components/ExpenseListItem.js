import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({id, description, amount = 0, createdAt = 0}) => (
    <li>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>Amount: {numeral(amount / 100).format('$0,00.00')}</p>
        <p>Created At {moment(createdAt).format('MMMM Do, YYYY')}</p>
    </li>
);

export default ExpenseListItem;