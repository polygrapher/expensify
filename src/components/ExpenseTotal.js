import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import {Link} from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expensesTotal';

export const ExpenseTotal = ({expensesCount, expensesTotalAmount}) => {
    const plural = 'expense' + (expensesCount > 1 ? 's' : '');
    const formattedTotal = numeral(expensesTotalAmount / 100).format('0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{expensesCount}</span> {plural} totalling <span>${formattedTotal}</span></h1>
                <div className="page-header__actions">
                    <Link to="/create" className="button">Add expense</Link>
                </div>
            </div>
        </div>
    );
}

const mapStateToProps = ({expenses, filters}) => {
    const selectedExpenses = selectExpenses(expenses, filters);
    return {
        expensesCount: selectedExpenses.length,
        expensesTotalAmount: expensesTotal(selectedExpenses)
    }
};

export default connect(mapStateToProps)(ExpenseTotal);