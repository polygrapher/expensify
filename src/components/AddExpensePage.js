import React from 'react';
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <h1>Add expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => bindActionCreators({startAddExpense}, dispatch);

export default connect(undefined, mapDispatchToProps)(AddExpensePage);