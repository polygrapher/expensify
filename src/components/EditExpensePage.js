import React from 'react';
import { bindActionCreators } from 'redux';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, startRemoveExpense, startEditExpense} from '../actions/expenses';
export class EditExpensePage extends React.Component {
    onSubmit = (updates) => {
        this.props.startEditExpense(this.props.expense.id, updates);
        this.props.history.push('/');
    }

    onRemoveButtonClick = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        // This may redirect before item was deleted in firebase
        this.props.history.push('/');

        // TODO figure out how to handle promise in jest.fn();
        // this.props.startRemoveExpense({id: this.props.expense.id}).then(() => {
        //     this.props.history.push('/');
        // }).catch((e) => {
        //     console.error('Failed to remove expense');
        // });
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                    <button className="button button--secondary" onClick={this.onRemoveButtonClick}>Remove Expense</button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id
    })
});

const mapDispatchToProps = (dispatch) => (bindActionCreators({editExpense, startRemoveExpense, startEditExpense}, dispatch));

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);