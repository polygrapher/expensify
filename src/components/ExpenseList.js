import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import ExpenseTotal from './ExpenseTotal';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = ({expenses = []}) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                expenses.length ? (
                    <ul className="list-body__items">
                        {expenses.map((expense) => <ExpenseListItem {...expense} key={expense.id} />)}
                    </ul>
                ) : (
                    <div className="list-item list-item--message">No expenses found</div>
                )
            }
        </div>
    </div>
);

const mapStateToProps = ({expenses, filters}) => ({
    expenses: selectExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);