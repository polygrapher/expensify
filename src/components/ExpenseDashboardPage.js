import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <div>This is a dashboard page</div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;