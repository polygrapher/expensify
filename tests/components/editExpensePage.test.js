import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../src/components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    startRemoveExpense = jest.fn();
    startEditExpense = jest.fn();
    history = {
        push: jest.fn()
    };
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[0]}
            history={history}
            startEditExpense={startEditExpense}
            startRemoveExpense={startRemoveExpense}
        />
    )
});

test('should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should handle startRemoveExpense expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[0].id});
});