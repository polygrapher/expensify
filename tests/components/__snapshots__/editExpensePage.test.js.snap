// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 10000,
          "createdAt": 0,
          "description": "Game",
          "id": "1",
          "note": "Note",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
